From 840d440c0fbb244ab5d981731521cc6fb037621c Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Wed, 23 Jun 2021 19:41:22 +0300
Subject: [PATCH] [SW-46517] greco: add axi drain support

This patch adds axi drain support for in ppboot for
secureboot flow.

Change-Id: I6f52d05d467b7a8817a41040b3fe3150b13d5997
---

diff --git a/arch/greco_arc/common.c b/arch/greco_arc/common.c
index 605f587..07a4edf 100644
--- a/arch/greco_arc/common.c
+++ b/arch/greco_arc/common.c
@@ -372,6 +372,18 @@
 	regmap_writel(0x1, rst_offset);
 }
 
+static void axi_enable_drain(void)
+{
+	u32 val;
+
+	/* Set timeout and enable drain */
+	regmap_writel(PSOC_AXI_DRAIN_TIMEOUT,
+			mmPSOC_GLOBAL_CONF_AXI_DRAIN_TIMEOUT);
+	val = regmap_readl(mmPSOC_GLOBAL_CONF_AXI_DRAIN_CTRL);
+	val |= (0x1 << PSOC_GLOBAL_CONF_AXI_DRAIN_CTRL_EN_SHIFT);
+	regmap_writel(val, mmPSOC_GLOBAL_CONF_AXI_DRAIN_CTRL);
+}
+
 /* initialize sram scrambling before preboot is downloaded */
 static void sram_scrambling_init(void)
 {
@@ -403,6 +415,8 @@
 {
 	init_arc();
 
+	axi_enable_drain();
+
 	sram_scrambling_init();
 }
 
diff --git a/arch/greco_arc/common.h b/arch/greco_arc/common.h
index 1385f27..667db54 100644
--- a/arch/greco_arc/common.h
+++ b/arch/greco_arc/common.h
@@ -51,4 +51,6 @@
 #define PCIE_WAIT_TIMEOUT_US	(5 * USEC_PER_SEC)
 #define PCIE_EXT_LD_TIMEOUT_US	576
 
+#define PSOC_AXI_DRAIN_TIMEOUT		0x3E8 /* 1000 cycles */
+
 #endif /* COMMON_GRECO_ARC_H_ */
